### ISSUE TITLE
Test Accuracy

### ISSUE THREAD

**Issue by pdcoded** [mozilla/DeepSpeech#502]
Hey @kdavis-mozilla , I have used a corpus similar like TED one and have created train,test and val folders inside data (wav and stm). Now I trained it on a huge batch of data and then the test results were decent. Now I changed the true transcriptions of test data inside stm folder to some random transcription . Now when I test it, the same test batch results were poor . What is wrong ? Does it tries to do ctc alignment against the ground truth output even while test ?

**Comment by kdavis-mozilla** [comment]
I'm not sure I understand your question. 

From what I understand you've created a new data set with train,test and validation data. You've trained on training data set, then you found the test results we acceptable. You then changed the transcriptions for the test data set to random strings and found the test results were then un-acceptable. Is this a fair description?

**Comment by pdcoded** [comment]
True . What can be the issue ?

**Comment by kdavis-mozilla** [comment]
There needs to be ground truth upon which the test results are evaluated. If you change the ground truth, then the test results will also change.

**Comment by pdcoded** [comment]
I am not sure if I agree. In an ideal way, we train the data and then use the trained model to predict the test output using the test audio files. Now after I get the test output, I can get the average WER across the entire batch against the ground truth output. This will be a true estimate of my accuracy

If changing the  ground truth output in test changes the test output ,then isnt the accuracy biased?

**Comment by reuben** [comment]
When you say the test results were poor, do you mean the network started started outputting incorrect text for the same audios it transcribed correctly before?

**Comment by kdavis-mozilla** [comment]
@reuben As far as I understand, @pdcoded correct me if I'm wrong, he trained a model with a training data set. He tested the model using a test data set, and it worked fine.

He then changed the text transcriptions for the test data set to random characters and ran the test on this new test data set. He then got bad results on the new test data set as ground truth was random text.

**Comment by pdcoded** [comment]
@reuben exactly. The network starting adapting to the new adversarial outputs that I randomly created and gave poor results. Now what I fear is during test, the ground truth output influencing the acoustic model using ctc loss?

**Comment by kdavis-mozilla** [comment]
@pdcoded The network doesn't train over the test data. So I don't see how changing the test data's transcription changes results. Am I misunderstanding what you're doing?

**Comment by kdavis-mozilla** [comment]
@pdcoded Do you simply mean that the WER for the test results changed when you changed the ground truth of the test set?

**Comment by pdcoded** [comment]
yeah, the outputs model predicted changed. And hence the WER, accuracy and loss metrics too . Can you try doing the same exercise on any small dataset?

**Comment by kdavis-mozilla** [comment]
I doubt the outputs the model predicted changed unless you continued to train on the training data set. Why do you think this is the case?

**Comment by pdcoded** [comment]
Okay please correct me I am wrong . I had trained my model to 300 epochs and then the last epoch displayed the test batch results smthng like "Test of Epoch 300 -" with src  and res. Now I  saved the last checkpoint. Changed the test true transcripts to random and re-instantianted ./bin/run-xxx.sh .Will that start re-training? Because what it now produced was something like  "Test of Epoch 300 -" but now with bad outputs. Please tell me,what am I doing wrong here .

**Comment by reuben** [comment]
The results are calculated as the difference between network output and ground truth. If the ground truth is garbage, the comparison is also garbage. What you're seeing is the expected behavior.

**Comment by kdavis-mozilla** [comment]
@pdcoded But this means you never saw the actually output from the network. You only saw the WER and as @reuben says this is expected behavior.

**Comment by pdcoded** [comment]
@reuben , @kdavis-mozilla so you guys mean to say "res" printed out is not the network output?

**Comment by reuben** [comment]
It is, but the report only prints the ten examples with the lowest loss. When the ground truths are random, the report will print ten random examples from the test set, which is probably why you're seeing worse results.

**Comment by kdavis-mozilla** [comment]
@pdcoded Could you post what you saw as the output of the network when doing this test?

In other words show us the output of the original 300 epoch run and show us the output of the run after you changed the ground truth of the test set.

I think this will help in clearing things up.

**Comment by pdcoded** [comment]
In the /bin/run-xxx.sh , I had actually given a bigger test_batch_size size of 100. Is there any way I can see all the results instead of lowest 10 losses. Can you point me to that part of the codebase.

Sure @kdavis-mozilla , I am sharing a screenshot after I changed to random
![screen shot 2017-04-07 at 12 16 50 am](https://cloud.githubusercontent.com/assets/9830931/24770547/e853d3f6-1b27-11e7-831d-86c685b23df2.jpg)

**Comment by pdcoded** [comment]
the test data that gave  res as  "o" and "ursiit" previously gave good results with less than 0.18 avg WER when the test transcripts were not imputed.

**Comment by reuben** [comment]
Like I said, previously you were looking at the 10 best results. Now you're looking at 10 more or less random results.

**Comment by pdcoded** [comment]
@reuben okay it is displaying 10 least loss outputs. But if the res is the original network output ,why is that changing? It should not change right. The loss would even then be high which is fine.

**Comment by reuben** [comment]
I doubt it's changing. What's changing is what specific examples are getting printed out.

**Comment by pdcoded** [comment]
okay , let me check that out by seeing the entire output set. Is there anyway I can see the output of the entire test batch instead of best 10 ?

**Comment by kdavis-mozilla** [comment]
@pdcoded It's the command line argument `report_count`

**Comment by pdcoded** [comment]
Okay got it !! Thanks :) I will try increasing that to the test batch itself. So if I am not wrong here, when it prints  "Test of Epoch 300 - WER: 7.993750, loss: 754.727600098, accuracy: 12.361112 " , the wer is the average WER over the entire test data set right ?

**Comment by kdavis-mozilla** [comment]
@pdcoded Test batch or test epoch? I think you want to increase to the test epoch size. 

Yes the results are averaged for the entire test data set.

**Comment by pdcoded** [comment]
Yeah doing that :) Really appreciate guys for addressing issues so promptly.  Appreciate this great open source collaboration and looking forward to resolving more critical issues :)
